docs/structured-streaming-programming-guide.md

```diff
@@ -0,0 +1,888 @@
+
+## Operations on streaming DataFrames/Datasets
+You can apply all kinds of streaming DataFrames/Datasets - ranging for untyped, SQL-like operations (e.g. select, where, groupBy), to typed RDD-like operations (e.g. map, filter, flatMap). See the SQL programming guide for more details. Let’s take a look at a few example operations that you can use.
+
+### Basic Operations - Selection, Projection, Aggregation
+Most of the common operations on DataFrame/Dataset operations are supported for streaming. The few operations that are not supported are discussed later in this section.
+
+<div class="codetabs">
+<div data-lang="scala"  markdown="1">
+
+{% highlight scala %}
+case class DeviceData(device: String, type: String, signal: Double, time: DateTime)
+
+val df: DataFrame = … // streaming DataFrame with IOT device data with schema { device: string, type: string, signal: double, time: string }
+val ds: Dataset[DeviceData] = df.as[DeviceData]    // streaming Dataset with IOT device data
+
+// Select the devices which have signal more than 10
+df.select("device").where("signal > 10")      // using untyped APIs   
+ds.filter(_.signal > 10).map(_.device)         // using typed APIs
+
+// Running count of the number of updates for each device type
+df.groupBy("type").count()                          // using untyped API
+
+// Running average signal for each device type
+Import org.apache.spark.sql.expressions.scalalang.typed._
+ds.groupByKey(_.type).agg(typed.avg(_.signal))    // using typed API
+{% endhighlight %}
+
+</div>
+<div data-lang="java"  markdown="1">
+
+</div>
+<div data-lang="python"  markdown="1">
+
+</div>
+</div>
+
+### Window Operations on Event Time
+Aggregations over a sliding window is very easy. Applying a sliding window is simply a special type of grouping - each time window is a group and each row can belong to multiple windows/groups. Hence aggregating over sliding windows is same as aggregating over this special type of grouping. This is illustrated below. Imagine the quick example is modified and the stream contains lines along with the time when the line was generated. Lets say instead of running word counts, we want to count windows within 10 minute windows, updating every 5 minutes. The result tables would look like the following. Note the window notation 12:00 - 12:10 means data that arrived after 12:00 but before 12:10. The counts are indexed by the combined group (window, cat)
+
+![Window Operations](img/structured-streaming-window.png)
+
+In code, you can use windows as the following. 
+
+<div class="codetabs">
+<div data-lang="scala"  markdown="1">
+
+{% highlight scala %}
+// Number of events in every 1 minute time windows
+df.groupBy(window(time, "1 minute"))
```

Comment:
keeps API and code snippets consistent across Java/scala/python. i agree its not the most optimized for scala users, but we are targeting getting convenience and consistency across languages over optimizing for one language.